File: api/src/main/java/com/sourceplusplus/api/model/metric/MetricType.java
package com.sourceplusplus.api.model.metric;

/**
 * Different metric order types core can be queried by.
 *
 * @version 0.3.1
 * @since 0.1.0
 * @author <a href="mailto:brandon@srcpl.us">Brandon Fergerson</a>
 */
public enum MetricType {

    Throughput_Average,
    ResponseTime_Average,
    ServiceLevelAgreement_Average,
    ResponseTime_99Percentile,
    ResponseTime_95Percentile,
    ResponseTime_90Percentile,
    ResponseTime_75Percentile,
    ResponseTime_50Percentile
}

File: protocol/src/commonMain/kotlin/com/sourceplusplus/protocol/artifact/metrics/MetricType.kt
package com.sourceplusplus.protocol.artifact.metrics

import kotlinx.serialization.Serializable

/**
 * todo: description.
 *
 * @since 0.1.0
 * @author [Brandon Fergerson](mailto:bfergerson@apache.org)
 */
@Serializable
enum class MetricType {
    Throughput_Average,
    ResponseTime_Average,
    ServiceLevelAgreement_Average,
    ResponseTime_99Percentile,
    ResponseTime_95Percentile,
    ResponseTime_90Percentile,
    ResponseTime_75Percentile,
    ResponseTime_50Percentile;

    val responseTimePercentile: Boolean
        get() = this == ResponseTime_99Percentile
                || this == ResponseTime_95Percentile
                || this == ResponseTime_90Percentile
                || this == ResponseTime_75Percentile
                || this == ResponseTime_50Percentile

    companion object {
        //todo: remove
        fun realValueOf(name: String): MetricType {
            return (values().find { it.name == name }
                ?: when (name) {
                    "endpoint_cpm" -> Throughput_Average
                    "endpoint_avg" -> ResponseTime_Average
                    "endpoint_sla" -> ServiceLevelAgreement_Average
                    "endpoint_percentile" -> ResponseTime_99Percentile
                    else -> throw UnsupportedOperationException(name)
                })
        }
    }
}
